README.md
Streaming API implementation of [MessagePack](https://msgpack.org/) binary serialization format - msgpack.

[![pub package](https://img.shields.io/pub/v/messagepack.svg)](https://pub.dartlang.org/packages/messagepack)
[![MIT License](https://img.shields.io/badge/License-BSD3-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

## The basics

`Packer` and `Unpacker` classes provide Streaming API for serializing and deserializing data.  
`Unpacker` also provide automatic unpacking `Map` and `List` by implicitly unpacking internal items.

## Streaming API Example - Simple

For the simplest data packaging to bytes packet like this: 
```
-------------------------------------------   
| version | userId | broadcast? | message |
-------------------------------------------
```

### Packer part

```dart
import 'package:messagepack/messagepack.dart';
```

```dart
final p = Packer();
p.packInt(1); // packet format version
p.packInt(222); // user id
p.packBool(false); // broadcast message to others
p.packString('hi'); // user message text
final bytes = p.takeBytes(); //Uint8List
yourFunctionSendToServer(bytes); //sends [1, 204, 222, 195, 162, 104, 105]
```
> `1` encodes to 1 byte with value 1.  
> `222` encodes to 2 bytes with values [204, 222] because 222 > 127.  
> `true` encodes to 1 byte with value 195.  
> `'hi'` encodes to 3 bytes with first byte containing str length info and other 2 bytes hold symbols values.  

> For more information, refer to the [msgpack](https://msgpack.org/) documentation.  

Streaming packing process not like the way we usually pack data in json, by specifying keys.  
The current example is more like how data is packed in TCP/IP frame structure.

### Unpacker part

```dart
final List<int> rawBytes = yourFunctionReceiveFromServer(); // receive List<int> bytes from server
final u = Unpacker.fromList(rawBytes);
final version = u.unpackInt();
final userId = u.unpackInt();
final broadcast = u.unpackBool();
final message = u.unpackString();

// check values in test
expect(version, equals(1));
expect(userId, equals(222));
expect(broadcast, equals(true));
expect(message, equals('hi'));

```

## Streaming packing and automatic/implicit unpacking example - complex

```dart
final p = Packer()
  ..packListLength(10)              //pack 10 different types items to list
  ..packInt(99)
  ..packBool(true)
  ..packString('hi')
  ..packNull()                      // explicitly pack null
  ..packString(null)                // implicitly any type can accept null 
  ..packBinary(<int>[104, 105])     // hi codes
  ..packListLength(2)               // pack 2 elements list ['elem1',3.14]
  ..packString('elem1')             // list[0]
  ..packDouble(3.14)                // list[1]
  ..packString('continue to pack other elements')
  ..packMapLength(1)                // map {'key1':false}
  ..packString('key1')              // map key
  ..packBool(false)                 // map value 
  ..packInt(9223372036854775807);   // next root list item (map ended)

final bytes = p.takeBytes();
final u = Unpacker(bytes);
// Unpack by the same sequential/streaming way 
// or implicitly/automatically
final l = u.unpackList(); //List<Object>
print(l);
// [99, true, hi, null, null, [104, 105], [elem1, 3.14], continue to pack other elements, {key1: false}, 9223372036854775807]
```

`List<Object>` items explicitly casted to corresponding types when using.

## NOTES

### Packing Maps and Lists
* firstly, pack Map or List header length 
* secondly, manually pack all items - that's all 

Only need put length header before packing items 

After packing all items no need to stop or finish or end this map / list

```dart
final list = ['i1','i2'];
final map = {'k1': 11, 'k2': 22};
final p = Packer();
p.packListLength(list.length);
list.forEach(p.packString);
p.packMapLength(map.length);
map.forEach((key, v) {
  p.packString(key);
  p.packInt(v);
});
final bytes = p.takeBytes();
```

### More examples

More examples can be found in:
* `test/messagepack_test.dart`
* `example/example.dart`

### Don't use Packer after calling .takeBytes()

Internally it cleans up underlying `_builder`, so further behaviour will be useless.  
Call `.takeBytes()` only once. It returns `Uint8List` of packed bytes.  
After that don't continue to use `Packer` instance (don't call `.packXXX()` methods and `.takeBytes()` anymore).  
Instead create new `Packer` instance.

### Motivation for creating this package

No other packages available that give streaming API for processing data 

## Contributing

If you have advice how to improve library code or making it lighter or blazingly faster - don't hesitate to open an issue or pull request!


CHANGELOG.md
## [0.2.0]

* Migrate to null safety
* Require Dart SDK 2.12 or later

## [0.1.2]

* Web support

## [0.1.1]

* New functions for _implicitly/automatically_ unpack `Map` and `List` items without using **Streaming API** directly.
* * `unpackList()`
* * `unpackMap()`
* Refactored tests and added more tests
* Improved documentation

## [0.1.0] 

* Few examples 
* More tests
* Improved documentation 

## [0.1.0-dev.1]  

* Pre release tested and functional streaming API


test/test.dart
import 'dart:convert';
import 'dart:math';


import 'package:messagepack/messagepack.dart';
import 'package:test/test.dart';

int packUnpackInt(int v, {bool negative = false}) {
  final p = Packer();
  if (negative)
    p.packInt(v);
  else
    p.packInt(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackInt();
}

double packUnpackDouble(double v) {
  final p = Packer();
  p.packDouble(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackDouble();
}

String packUnpackString(String v) {
  final p = Packer();
  p.packString(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackString();
}

final ints = [
  9223372036854775807,
  1581367842777,
  0,
  1581367842,
  127,
  5,
  128,
  null,
  128,
];

final negativeInts = ints.map((e) => e == null ? null : -1 * e).toList()
  ..add(-9223372036854775808);

final doubles = [
  0.0,
  null,
  1.1,
  null,
  null,
  12345678.12345668,
  null,
  -1.1,
  2.56,
  -34.567890,
  -23456.54,
  null,
  -1.0
];

final strs = [
  'hi',
  null,
  null,
  'Повседневная практика показывает, что постоянное информационно-пропагандистское обеспечение нашей деятельности в значительной степени обуславливает создание модели развития. Не следует, однако забывать, что реализация намеченных плановых заданий требуют от нас анализа форм развития. Не следует, однако забывать, что новая модель организационной деятельности требуют определения и уточнения модели развития. Равным образом постоянный количественный рост и сфера нашей активности требуют от нас анализа системы обучения кадров, соответствует насущным потребностям.',
  'mediul length string',
  '12345678901234567890123456789012345678901234567890',
  '123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890',
  null,
  'str containing  кириллицу',
  'hi'
];

void main() {
  test('int sequentially increasing [independent]', () {
    for (int v = -65536; v < 65536; v++) {
      final r = packUnpackInt(v, negative: v < 0);
      expect(r, equals(v));
    }
  });

  test('Positive int [independent]', () {
    for (final v in ints) expect(packUnpackInt(v), equals(v));
  });

  test('Positive int [dependent]', () {
    final p = Packer();
    ints.forEach(p.packInt);
    final u = Unpacker(p.takeBytes());
    for (final v in ints) expect(u.unpackInt(), equals(v));
  });

  test('Negative int [independent]', () {
    for (final v in negativeInts)
      expect(packUnpackInt(v, negative: true), equals(v));
  });

  test('Negative int [dependent]', () {
    final p = Packer();
    negativeInts.forEach(p.packInt);
    final u = Unpacker(p.takeBytes());
    for (final v in negativeInts) expect(u.unpackInt(), equals(v));
  });

  test('Double [independent]', () {
    for (final v in doubles) expect(packUnpackDouble(v), equals(v));
  });

  test('Double [dependent]', () {
    final p = Packer();
    doubles.forEach(p.packDouble);
    final u = Unpacker(p.takeBytes());
    for (final v in doubles) expect(u.unpackDouble(), equals(v));
  });

  test('String [independent]', () {
    for (final v in strs) expect(packUnpackString(v), equals(v));
  });

  test('String [dependent]', () {
    final p = Packer();
    strs.forEach(p.packString);
    final u = Unpacker(p.takeBytes());
    for (final v in strs) expect(u.unpackString(), equals(v));
  });

  List<MapEntry<int, dynamic>> randomiseData() {
    final rands = <MapEntry<int, dynamic>>[];
    MapEntry<int, dynamic> me(int k, dynamic v) => MapEntry<int, dynamic>(k, v);
    final r = Random();

    dynamic nullOr(dynamic transform(int v)) {
      return r.nextInt(10) == 0 ? null : transform(r.nextInt(4294967296));
    }

    for (int i = 0; i < 200; i++) {
      final t = r.nextInt(5);
      if (t == 0) rands.add(me(t, null));
      if (t == 1) rands.add(me(t, nullOr((v) => v % 3 == 0)));
      if (t == 2) rands.add(me(t, nullOr((v) => v)));
      if (t == 3) rands.add(me(t, nullOr((v) => v * -1)));
      if (t == 4)
        rands.add(me(t,
            nullOr((v) => r.nextDouble() * v * (r.nextInt(2) == 1 ? -1 : 1))));
      if (t == 5) rands.add(me(t, nullOr((v) => strs[r.nextInt(strs.length)])));
    }
    return rands;
  }

  void pack(Packer p, MapEntry<int, dynamic> e) {
    if (e.key == 0) p.packNull();
    if (e.key == 1) p.packBool(e.value as bool);
    if (e.key == 2) p.packInt(e.value as int);
    if (e.key == 3) p.packInt(e.value as int);
    if (e.key == 4) p.packDouble(e.value as double);
    if (e.key == 5) p.packString(e.value as String);
  }

  dynamic unpack(Unpacker u, int type) {
    if (type == 0) return unpack(u, Random().nextInt(4) + 1);
    if (type == 1) return u.unpackBool();
    if (type == 2) return u.unpackInt();
    if (type == 3) return u.unpackInt();
    if (type == 4) return u.unpackDouble();
    if (type == 5) return u.unpackString();
  }

  test('Random different [independent]', () {
    final l = randomiseData();
    for (final e in l) {
      final p = Packer();
      pack(p, e);
      final u = Unpacker(p.takeBytes());
      expect(unpack(u, e.key), equals(e.value));
    }
  });

  test('Random different [dependent]', () {
    final l = randomiseData();
    final p = Packer();
    for (final e in l) {
      pack(p, e);
    }
    final u = Unpacker(p.takeBytes());
    for (int i = 0; i < l.length; i++) {
      expect(unpack(u, l[i].key), equals(l[i].value));
    }
  });

  test('Iterable [dependent]', () {
    final list = [
      'value1 short',
      'value2 medium medium',
      'value3 with кириллица кириллица кириллица',
      'value4 with long long text that so long long long long long long long long long long',
      'value5',
      'v6'
    ];
    final p = Packer();
    p.packNull();
    p.packIterableLength(null);
    p.packIterableLength(0);
    p.packIterableLength(list.length);
    list.forEach(p.packString);
    p.packBool(true);
    p.packIterableLength(list.length);
    list.forEach(p.packString);
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackIterableLength(), equals(0));
    expect(u.unpackIterableLength(), equals(0));
    expect(u.unpackIterableLength(), equals(0));
    final length = u.unpackIterableLength();
    for (int i = 0; i < length; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    expect(u.unpackBool(), equals(true));
    final length2 = u.unpackIterableLength();
    for (int i = 0; i < length2; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    expect(u.unpackInt(), equals(3));
  });

  test('Iterable different types [dependent]', () {
    final list = [
      -65999,
      -34855,
      -128,
      -31,
      -1,
      0,
      null,
      124,
      1581367842,
      1581367842777,
      null,
      'short string',
      'string with medium length',
      true,
      null,
      false,
      -3.14,
      0.5,
      36.6,
      null,
      100.0,
    ];
    final p = Packer();
    p.packIterableLength(list.length);
    for (int i = 0; i < 10; i++) {
      p.packInt(list[i] as int);
    }
    for (int i = 10; i < 13; i++) {
      p.packString(list[i] as String);
    }
    for (int i = 13; i < 16; i++) {
      p.packBool(list[i] as bool);
    }
    for (int i = 16; i < 21; i++) {
      p.packDouble(list[i] as double);
    }
    final u = Unpacker(p.takeBytes());
    expect(u.unpackIterableLength(), equals(list.length));
    for (int i = 0; i < 10; i++) {
      expect(u.unpackInt(), equals(list[i]));
    }
    for (int i = 10; i < 13; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    for (int i = 13; i < 16; i++) {
      expect(u.unpackBool(), equals(list[i]));
    }
    for (int i = 16; i < 21; i++) {
      expect(u.unpackDouble(), equals(list[i]));
    }
  });

  test('Map [dependent]', () {
    final map = {
      1: 11,
      2: 22,
      3: 33,
    };
    final p = Packer();
    p.packNull();
    p.packMapLength(null);
    p.packMapLength(0);
    p.packMapLength(map.length);
    for (int i = 0; i < map.length; i++) {
      p.packInt(i);
      p.packInt(map[i]);
    }
    p.packBool(true);
    p.packMapLength(map.length);
    for (int i = 0; i < map.length; i++) {
      p.packInt(i);
      p.packInt(map[i]);
    }
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackMapLength(), equals(0));
    expect(u.unpackMapLength(), equals(0));
    expect(u.unpackMapLength(), equals(0));
    final length = u.unpackMapLength();
    for (int i = 0; i < length; i++) {
      expect(u.unpackInt(), equals(i));
      expect(u.unpackInt(), equals(map[i]));
    }
    expect(u.unpackBool(), equals(true));
    final length2 = u.unpackMapLength();
    for (int i = 0; i < length2; i++) {
      expect(u.unpackInt(), equals(i));
      expect(u.unpackInt(), equals(map[i]));
    }
    expect(u.unpackInt(), equals(3));
  });

  test('Binary [dependent]', () {
    final bytes1 = utf8.encode('hi');
    final bytes2 = utf8.encode(
        'Значимость этих проблем настолько очевидна, что укрепление и развитие структуры представляет собой интересный эксперимент проверки систем массового участия. Задача организации, в особенности же постоянное информационно-пропагандистское обеспечение нашей деятельности обеспечивает широкому кругу (специалистов) участие в формировании систем массового участия. Товарищи! постоянный количественный рост и сфера нашей активности требуют определения и уточнения соответствующий условий активизации. Задача организации, в особенности же консультация с широким активом влечет за собой процесс внедрения и модернизации форм развития. Равным образом начало повседневной работы по формированию позиции позволяет оценить значение соответствующий условий активизации.');
    final empty = <int>[];
    final p = Packer();
    p.packBinary(null);
    p.packBinary(empty);
    p.packBinary(bytes1);
    p.packBool(true);
    p.packBinary(bytes2);
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackBinary(), equals(empty));
    expect(u.unpackBinary(), equals(empty));
    expect(u.unpackBinary(), equals(bytes1));
    expect(u.unpackBool(), equals(true));
    expect(u.unpackBinary(), equals(bytes2));
    expect(u.unpackInt(), equals(3));
  });

  test('Manual int example [dependent]', () {
    final p = Packer();
    p.packInt(1);
    p.packInt(2);
    final bytes = p.takeBytes();
    //send bytes to server
    //receive bytes from server
    final u = Unpacker(bytes);
    final n1 = u.unpackInt();
    final n2 = u.unpackInt();
    expect(n1, equals(1));
    expect(n2, equals(2));
  });

  test('Map Iterable example [dependent]', () {
    final list = ['i1','i2'];
    final map = {'k1': 11, 'k2': 22};
    final p = Packer();
    p.packIterableLength(list.length);
    list.forEach(p.packString);
    p.packMapLength(map.length);
    map.forEach((key, v) {
      p.packString(key);
      p.packInt(v);
    });
    final bytes = p.takeBytes();
  });
}


test/msgpack_test.dart
import 'dart:convert';
import 'dart:math';


import 'package:msgpack/msgpack.dart';
import 'package:test/test.dart';

int packUnpackInt(int v, {bool negative = false}) {
  final p = Packer();
  if (negative)
    p.packInt(v);
  else
    p.packInt(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackInt();
}

double packUnpackDouble(double v) {
  final p = Packer();
  p.packDouble(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackDouble();
}

String packUnpackString(String v) {
  final p = Packer();
  p.packString(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackString();
}

final ints = [
  9223372036854775807,
  1581367842777,
  0,
  1581367842,
  127,
  5,
  128,
  null,
  128,
];

final negativeInts = ints.map((e) => e == null ? null : -1 * e).toList()
  ..add(-9223372036854775808);

final doubles = [
  0.0,
  null,
  1.1,
  null,
  null,
  12345678.12345668,
  null,
  -1.1,
  2.56,
  -34.567890,
  -23456.54,
  null,
  -1.0
];

final strs = [
  'hi',
  null,
  null,
  'Повседневная практика показывает, что постоянное информационно-пропагандистское обеспечение нашей деятельности в значительной степени обуславливает создание модели развития. Не следует, однако забывать, что реализация намеченных плановых заданий требуют от нас анализа форм развития. Не следует, однако забывать, что новая модель организационной деятельности требуют определения и уточнения модели развития. Равным образом постоянный количественный рост и сфера нашей активности требуют от нас анализа системы обучения кадров, соответствует насущным потребностям.',
  'mediul length string',
  '12345678901234567890123456789012345678901234567890',
  '123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890',
  null,
  'str containing  кириллицу',
  'hi'
];

void main() {
  test('int sequentially increasing [independent]', () {
    for (int v = -65536; v < 65536; v++) {
      final r = packUnpackInt(v, negative: v < 0);
      expect(r, equals(v));
    }
  });

  test('Positive int [independent]', () {
    for (final v in ints) expect(packUnpackInt(v), equals(v));
  });

  test('Positive int [dependent]', () {
    final p = Packer();
    ints.forEach(p.packInt);
    final u = Unpacker(p.takeBytes());
    for (final v in ints) expect(u.unpackInt(), equals(v));
  });

  test('Negative int [independent]', () {
    for (final v in negativeInts)
      expect(packUnpackInt(v, negative: true), equals(v));
  });

  test('Negative int [dependent]', () {
    final p = Packer();
    negativeInts.forEach(p.packInt);
    final u = Unpacker(p.takeBytes());
    for (final v in negativeInts) expect(u.unpackInt(), equals(v));
  });

  test('Double [independent]', () {
    for (final v in doubles) expect(packUnpackDouble(v), equals(v));
  });

  test('Double [dependent]', () {
    final p = Packer();
    doubles.forEach(p.packDouble);
    final u = Unpacker(p.takeBytes());
    for (final v in doubles) expect(u.unpackDouble(), equals(v));
  });

  test('String [independent]', () {
    for (final v in strs) expect(packUnpackString(v), equals(v));
  });

  test('String [dependent]', () {
    final p = Packer();
    strs.forEach(p.packString);
    final u = Unpacker(p.takeBytes());
    for (final v in strs) expect(u.unpackString(), equals(v));
  });

  List<MapEntry<int, dynamic>> randomiseData() {
    final rands = <MapEntry<int, dynamic>>[];
    MapEntry<int, dynamic> me(int k, dynamic v) => MapEntry<int, dynamic>(k, v);
    final r = Random();

    dynamic nullOr(dynamic transform(int v)) {
      return r.nextInt(10) == 0 ? null : transform(r.nextInt(4294967296));
    }

    for (int i = 0; i < 200; i++) {
      final t = r.nextInt(5);
      if (t == 0) rands.add(me(t, null));
      if (t == 1) rands.add(me(t, nullOr((v) => v % 3 == 0)));
      if (t == 2) rands.add(me(t, nullOr((v) => v)));
      if (t == 3) rands.add(me(t, nullOr((v) => v * -1)));
      if (t == 4)
        rands.add(me(t,
            nullOr((v) => r.nextDouble() * v * (r.nextInt(2) == 1 ? -1 : 1))));
      if (t == 5) rands.add(me(t, nullOr((v) => strs[r.nextInt(strs.length)])));
    }
    return rands;
  }

  void pack(Packer p, MapEntry<int, dynamic> e) {
    if (e.key == 0) p.packNull();
    if (e.key == 1) p.packBool(e.value as bool);
    if (e.key == 2) p.packInt(e.value as int);
    if (e.key == 3) p.packInt(e.value as int);
    if (e.key == 4) p.packDouble(e.value as double);
    if (e.key == 5) p.packString(e.value as String);
  }

  dynamic unpack(Unpacker u, int type) {
    if (type == 0) return unpack(u, Random().nextInt(4) + 1);
    if (type == 1) return u.unpackBool();
    if (type == 2) return u.unpackInt();
    if (type == 3) return u.unpackInt();
    if (type == 4) return u.unpackDouble();
    if (type == 5) return u.unpackString();
  }

  test('Random different [independent]', () {
    final l = randomiseData();
    for (final e in l) {
      final p = Packer();
      pack(p, e);
      final u = Unpacker(p.takeBytes());
      expect(unpack(u, e.key), equals(e.value));
    }
  });

  test('Random different [dependent]', () {
    final l = randomiseData();
    final p = Packer();
    for (final e in l) {
      pack(p, e);
    }
    final u = Unpacker(p.takeBytes());
    for (int i = 0; i < l.length; i++) {
      expect(unpack(u, l[i].key), equals(l[i].value));
    }
  });

  test('Iterable [dependent]', () {
    final list = [
      'value1 short',
      'value2 medium medium',
      'value3 with кириллица кириллица кириллица',
      'value4 with long long text that so long long long long long long long long long long',
      'value5',
      'v6'
    ];
    final p = Packer();
    p.packNull();
    p.packIterableLength(null);
    p.packIterableLength(0);
    p.packIterableLength(list.length);
    list.forEach(p.packString);
    p.packBool(true);
    p.packIterableLength(list.length);
    list.forEach(p.packString);
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackIterableLength(), equals(0));
    expect(u.unpackIterableLength(), equals(0));
    expect(u.unpackIterableLength(), equals(0));
    final length = u.unpackIterableLength();
    for (int i = 0; i < length; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    expect(u.unpackBool(), equals(true));
    final length2 = u.unpackIterableLength();
    for (int i = 0; i < length2; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    expect(u.unpackInt(), equals(3));
  });

  test('Iterable different types [dependent]', () {
    final list = [
      -65999,
      -34855,
      -128,
      -31,
      -1,
      0,
      null,
      124,
      1581367842,
      1581367842777,
      null,
      'short string',
      'string with medium length',
      true,
      null,
      false,
      -3.14,
      0.5,
      36.6,
      null,
      100.0,
    ];
    final p = Packer();
    p.packIterableLength(list.length);
    for (int i = 0; i < 10; i++) {
      p.packInt(list[i] as int);
    }
    for (int i = 10; i < 13; i++) {
      p.packString(list[i] as String);
    }
    for (int i = 13; i < 16; i++) {
      p.packBool(list[i] as bool);
    }
    for (int i = 16; i < 21; i++) {
      p.packDouble(list[i] as double);
    }
    final u = Unpacker(p.takeBytes());
    expect(u.unpackIterableLength(), equals(list.length));
    for (int i = 0; i < 10; i++) {
      expect(u.unpackInt(), equals(list[i]));
    }
    for (int i = 10; i < 13; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    for (int i = 13; i < 16; i++) {
      expect(u.unpackBool(), equals(list[i]));
    }
    for (int i = 16; i < 21; i++) {
      expect(u.unpackDouble(), equals(list[i]));
    }
  });

  test('Map [dependent]', () {
    final map = {
      1: 11,
      2: 22,
      3: 33,
    };
    final p = Packer();
    p.packNull();
    p.packMapLength(null);
    p.packMapLength(0);
    p.packMapLength(map.length);
    for (int i = 0; i < map.length; i++) {
      p.packInt(i);
      p.packInt(map[i]);
    }
    p.packBool(true);
    p.packMapLength(map.length);
    for (int i = 0; i < map.length; i++) {
      p.packInt(i);
      p.packInt(map[i]);
    }
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackMapLength(), equals(0));
    expect(u.unpackMapLength(), equals(0));
    expect(u.unpackMapLength(), equals(0));
    final length = u.unpackMapLength();
    for (int i = 0; i < length; i++) {
      expect(u.unpackInt(), equals(i));
      expect(u.unpackInt(), equals(map[i]));
    }
    expect(u.unpackBool(), equals(true));
    final length2 = u.unpackMapLength();
    for (int i = 0; i < length2; i++) {
      expect(u.unpackInt(), equals(i));
      expect(u.unpackInt(), equals(map[i]));
    }
    expect(u.unpackInt(), equals(3));
  });

  test('Binary [dependent]', () {
    final bytes1 = utf8.encode('hi');
    final bytes2 = utf8.encode(
        'Значимость этих проблем настолько очевидна, что укрепление и развитие структуры представляет собой интересный эксперимент проверки систем массового участия. Задача организации, в особенности же постоянное информационно-пропагандистское обеспечение нашей деятельности обеспечивает широкому кругу (специалистов) участие в формировании систем массового участия. Товарищи! постоянный количественный рост и сфера нашей активности требуют определения и уточнения соответствующий условий активизации. Задача организации, в особенности же консультация с широким активом влечет за собой процесс внедрения и модернизации форм развития. Равным образом начало повседневной работы по формированию позиции позволяет оценить значение соответствующий условий активизации.');
    final empty = <int>[];
    final p = Packer();
    p.packBinary(null);
    p.packBinary(empty);
    p.packBinary(bytes1);
    p.packBool(true);
    p.packBinary(bytes2);
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackBinary(), equals(empty));
    expect(u.unpackBinary(), equals(empty));
    expect(u.unpackBinary(), equals(bytes1));
    expect(u.unpackBool(), equals(true));
    expect(u.unpackBinary(), equals(bytes2));
    expect(u.unpackInt(), equals(3));
  });

  test('Manual int example [dependent]', () {
    final p = Packer();
    p.packInt(1);
    p.packInt(2);
    final bytes = p.takeBytes();
    //send bytes to server
    //receive bytes from server
    final u = Unpacker(bytes);
    final n1 = u.unpackInt();
    final n2 = u.unpackInt();
    expect(n1, equals(1));
    expect(n2, equals(2));
  });

  test('Map Iterable example [dependent]', () {
    final list = ['i1','i2'];
    final map = {'k1': 11, 'k2': 22};
    final p = Packer();
    p.packIterableLength(list.length);
    list.forEach(p.packString);
    p.packMapLength(map.length);
    map.forEach((key, v) {
      p.packString(key);
      p.packInt(v);
    });
    final bytes = p.takeBytes();
  });
}


lib/src/bytesbuilder.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// @dart = 2.6
import 'dart:typed_data';

/// Builds a list of bytes, allowing bytes and lists of bytes to be added at the
/// end.
///
/// Used to efficiently collect bytes and lists of bytes.
abstract class BytesBuilder {
  /// Construct a new empty [BytesBuilder].
  ///
  /// If [copy] is true, the data is always copied when added to the list. If
  /// it [copy] is false, the data is only copied if needed. That means that if
  /// the lists are changed after added to the [BytesBuilder], it may effect the
  /// output. Default is `true`.
  factory BytesBuilder({bool copy: true}) {
    if (copy) {
      return _CopyingBytesBuilder();
    } else {
      return _BytesBuilder();
    }
  }

  /// Appends [bytes] to the current contents of the builder.
  ///
  /// Each value of [bytes] will be bit-representation truncated to the range
  /// 0 .. 255.
  void add(List<int> bytes);

  /// Append [byte] to the current contents of the builder.
  ///
  /// The [byte] will be bit-representation truncated to the range 0 .. 255.
  void addByte(int byte);

  /// Returns the contents of `this` and clears `this`.
  ///
  /// The list returned is a view of the internal buffer, limited to the
  /// [length].
  Uint8List takeBytes();

  /// Returns a copy of the current contents of the builder.
  ///
  /// Leaves the contents of the builder intact.
  Uint8List toBytes();

  /// The number of bytes in the builder.
  int get length;

  /// Returns `true` if the buffer is empty.
  bool get isEmpty;

  /// Returns `true` if the buffer is not empty.
  bool get isNotEmpty;

  /// Clear the contents of the builder.
  void clear();
}

class _CopyingBytesBuilder implements BytesBuilder {
  // Start with 1024 bytes.
  static const int _initSize = 1024;

  // Safe for reuse because a fixed-length empty list is immutable.
  static final _emptyList = Uint8List(0);

  int _length = 0;
  Uint8List _buffer;

  // ignore: sort_constructors_first
  _CopyingBytesBuilder([int initialCapacity = 0])
      : _buffer = (initialCapacity <= 0)
      ? _emptyList
      : Uint8List(_pow2roundup(initialCapacity));

  @override
  void add(List<int> bytes) {
    final int bytesLength = bytes.length;
    if (bytesLength == 0) return;
    final int required = _length + bytesLength;
    if (_buffer.length < required) {
      _grow(required);
    }
    assert(_buffer.length >= required);
    if (bytes is Uint8List) {
      _buffer.setRange(_length, required, bytes);
    } else {
      for (int i = 0; i < bytesLength; i++) {
        _buffer[_length + i] = bytes[i];
      }
    }
    _length = required;
  }

  @override
  void addByte(int byte) {
    if (_buffer.length == _length) {
      // The grow algorithm always at least doubles.
      // If we added one to _length it would quadruple unnecessarily.
      _grow(_length);
    }
    assert(_buffer.length > _length);
    _buffer[_length] = byte;
    _length++;
  }

  void _grow(int required) {
    // We will create a list in the range of 2-4 times larger than
    // required.
    int newSize = required * 2;
    if (newSize < _initSize) {
      newSize = _initSize;
    } else {
      newSize = _pow2roundup(newSize);
    }
    var newBuffer = Uint8List(newSize);
    newBuffer.setRange(0, _buffer.length, _buffer);
    _buffer = newBuffer;
  }

  @override
  Uint8List takeBytes() {
    if (_length == 0) return _emptyList;
    var buffer =
    Uint8List.view(_buffer.buffer, _buffer.offsetInBytes, _length);
    clear();
    return buffer;
  }

  @override
  Uint8List toBytes() {
    if (_length == 0) return _emptyList;
    return Uint8List.fromList(
        Uint8List.view(_buffer.buffer, _buffer.offsetInBytes, _length));
  }

  @override
  int get length => _length;

  @override
  bool get isEmpty => _length == 0;

  @override
  bool get isNotEmpty => _length != 0;

  void clear() {
    _length = 0;
    _buffer = _emptyList;
  }

  static int _pow2roundup(int x) {
    assert(x > 0);
    --x;
    x |= x >> 1;
    x |= x >> 2;
    x |= x >> 4;
    x |= x >> 8;
    x |= x >> 16;
    return x + 1;
  }
}

class _BytesBuilder implements BytesBuilder {
  int _length = 0;
  final List<Uint8List> _chunks = [];

  @override
  void add(List<int> bytes) {
    Uint8List typedBytes;
    if (bytes is Uint8List) {
      typedBytes = bytes;
    } else {
      typedBytes = Uint8List.fromList(bytes);
    }
    _chunks.add(typedBytes);
    _length += typedBytes.length;
  }

  @override
  void addByte(int byte) {
    _chunks.add(Uint8List(1)..[0] = byte);
    _length++;
  }

  @override
  Uint8List takeBytes() {
    if (_length == 0) return _CopyingBytesBuilder._emptyList;
    if (_chunks.length == 1) {
      final buffer = _chunks[0];
      clear();
      return buffer;
    }
    final buffer = Uint8List(_length);
    int offset = 0;
    // ignore: prefer_final_in_for_each
    for (var chunk in _chunks) {
      buffer.setRange(offset, offset + chunk.length, chunk);
      offset += chunk.length;
    }
    clear();
    return buffer;
  }

  @override
  Uint8List toBytes() {
    if (_length == 0) return _CopyingBytesBuilder._emptyList;
    final buffer = Uint8List(_length);
    int offset = 0;
    // ignore: prefer_final_in_for_each
    for (var chunk in _chunks) {
      buffer.setRange(offset, offset + chunk.length, chunk);
      offset += chunk.length;
    }
    return buffer;
  }

  @override
  int get length => _length;

  @override
  bool get isEmpty => _length == 0;

  @override
  bool get isNotEmpty => _length != 0;

  @override
  void clear() {
    _length = 0;
    _chunks.clear();
  }
}


example/example.dart
import 'package:messagepack/messagepack.dart';

void simple() {
  final p = Packer();
  p.packInt(1);
  p.packInt(2);
  final bytes = p.takeBytes(); //Uint8List
  print(bytes);

  final u = Unpacker(bytes);
  final n1 = u.unpackInt();
  final n2 = u.unpackInt();
  print('unpacked n1=$n1 n2=$n2');
}

void differentTypesSimple() {
  final p = Packer();
  p.packInt(1);
  p.packBool(true);
  final bytes = p.takeBytes(); //Uint8List
  print(bytes);

  final u = Unpacker(bytes);
  print(u.unpackInt());
  print(u.unpackBool());
}

void iterableAndMap() {
  final list = ['i1', 'i2'];
  final map = {'k1': 11, 'k2': 22};
  final p = Packer();
  p.packListLength(list.length);
  list.forEach(p.packString);
  p.packMapLength(map.length);
  map.forEach((key, v) {
    p.packString(key);
    p.packInt(v);
  });
  final bytes = p.takeBytes();

  final u = Unpacker(bytes);
  final listLength = u.unpackListLength();
  for (int i = 0; i < listLength; i++) {
    print(u.unpackString());
  }
  final mapLength = u.unpackMapLength();
  for (int i = 0; i < listLength; i++) {
    print(u.unpackString());
    print(u.unpackInt());
  }
}

void differentTypesComplex() {
  final p = Packer()
    ..packInt(99)
    ..packBool(true)
    ..packString('hi')
    ..packNull()
    ..packString(null)
    ..packBinary(<int>[104, 105]) // hi codes
    ..packListLength(2) // pack 2 elements list ['elem1',3.14]
    ..packString('elem1')
    ..packDouble(3.14)
    ..packString('continue to pack other elements')
    ..packMapLength(2) //map {'key1':false, 'key2',3.14}
    ..packString('key1') //pack key1
    ..packBool(false) //pack value1
    ..packString('key12') //pack key1
    ..packDouble(3.13); //pack value1

  final bytes = p.takeBytes();
  final u = Unpacker(bytes);
  //Unpack the same sequential/streaming way
}
